File: base/src/main/java/mf/asciitext/fonts/AppFont.java
package mf.asciitext.fonts;

public abstract class AppFont {

    protected String id;
    protected String name;
    private boolean premium;
    private boolean favorite = false;
    protected String text;

    public AppFont(String id, String name, boolean premium) {
        this.id = id;
        this.name = name;
        this.premium = premium;
    }

    public String GetId() {
        return id;
    }

    public String GetName() {
        return name;
    }

    public String GetText() {
        return text;
    }

    public boolean isReversed() {
        return false;
    }

    public String GetStyledName() {
        return encode(name);
    }

    public boolean IsPremium() {
        return premium;
    }

    public boolean IsFavorite() {
        return favorite;
    }

    public void SetFavorite(boolean value) {
        favorite = value;
    }

    boolean isEmpty(String text) {
        boolean isEmpty = (text == null || text.length() == 0);
        if (isEmpty) {
            SetText(name);
        }
        return isEmpty;
    }

    public void SetText(String text) {
        this.text = encode(text);
    }

    public abstract String encode(String text);
}

File: base/src/main/java/mf/asciitext/fonts/AppFont.kt
package mf.asciitext.fonts

abstract class AppFont(
    protected var id: String,
    protected var name: String,
    private var _enabled: Boolean,
    private var _priority: Int,
    /**
     * This property should be set true for font styles
     * that typed in RIGHT -> LEFT direction
     */
    protected var reverse: Boolean = false,
    /**
     * This property should be set true for fonts where
     * encoding depends on a sequence of preceding characters
     */
    private var sequenceAware: Boolean = false
) {
    val fontId: String
        get() = id

    val styledName: String
        get() = encode(name)!!

    val isReversed: Boolean
        get() = reverse

    var isEnabled: Boolean
        get() = _enabled
        set(value) {
            _enabled = value
        }
    var priority: Int
        get() = _priority
        set(value) {
            _priority = value
        }

    val isSequenceAware: Boolean
        get() = sequenceAware

    fun isEmpty(text: String?): Boolean {
        return text == null || text.isEmpty()
    }

    abstract fun encode(text: String?, sequence: CharSequence? = null): String?
}